core/capability-negotiation.md

```diff
@@ -0,0 +1,457 @@
+subcommands below to list and request capabilities.
+
+Upon connecting to the server, the client attempts to start the capability negotiation
+process, as well as sending the standard `NICK`/`USER` commands (to complete registration
+if the server doesn't support capability negotiation).
+
+Upon connecting to the IRC server, clients SHOULD send one of the following messages:
+
+- `CAP LS [version]` to discover the available capabilities on the server.
+- `CAP REQ` to blindly request a particular set of capabilities.
+
+Following this, the client MUST send the standard `NICK` and `USER` IRC commands.
+
+Upon receiving either a `CAP LS` or `CAP REQ` command during connection registration, the
+server MUST not complete registration until the client sends a `CAP END` command to
+indicate that capability negotiation has ended. This allows clients to request their
+desired capabilities before completing registration.
+
+Once capability negotiation has completed with a client-send `CAP END` command, registration
+continues as normal.
+
+With the above registration process, clients will either recieve a `CAP` message indicating
+that registration is paused while capability negotiation happens, or registration will
+complete immediately, indicating that the server doesn't support capability negotiation.
+
+Example with capability negotiation:
+
+    Client: CAP LS 302
+    Client: NICK dan
+    Client: USER d * 0 :This is a really good name
+    Server: CAP * LS :multi-prefix sasl
+    Client: CAP REQ :multi-prefix
+    Server: CAP * ACK multi-prefix
+    Client: CAP END
+    Server: 001 dan :Welcome to the Internet Relay Network dan
+    ...
+
+Example with capability negotiation, but where the client recognises no advertised caps: 
+
+    Client: CAP LS 302
+    Client: NICK dan
+    Client: USER d * 0 :This is a really good name
+    Server: CAP * LS :draft/example-1 draft/example-2
+    Client: CAP END
+    Server: 001 dan :Welcome to the Internet Relay Network dan
+    ...
+
+Example where the server doesn't support capability negotiation:
+
+    Client: CAP LS 302
+    Client: NICK dan
+    Client: USER d * 0 :This is a really good name
+    Server: 001 dan :Welcome to the Internet Relay Network dan
+    ...
+
+Example where the client doesn't support capability negotiation:
+
+    Client: NICK dan
+    Client: USER d * 0 :This is a really good name
+    Server: 001 dan :Welcome to the Internet Relay Network dan
+    ...
+
+## The CAP command
+
+The client capability negotiation extension is implemented by the addition of one command
+with several subcommands.  The command added is named `CAP`.  `CAP` takes a single
+required subcommand, optionally followed by a single parameter of space-separated capability
+identifiers.  Each capability in the list may be preceded by a capability modifier.
+
+The subcommands for `CAP` are: `LS`, `LIST`, `REQ`, `ACK`, `NAK`, and `END`.
```

Comment:
Add NEW and DEL and make this a table indicating which are client sent and which are server sent, maybe a column indicating what they're sent in response to. e.g.

| Subcommand | Client sent | Server sent (triggered by) |
|------------|-------------|----------------------------|
| LS         | *           | * (LS)                     |
| LIST       | *           | * (LIST)                   |
| REQ        | *           |                            |
| ACK        |             | * (REQ)                    |
| NAK        |             | * (REQ)                    |
| END        | *           |                            |
| NEW (302)        |             | *                          |
| DEL (302)        |             | *                          |

Reply:
Note to self, look at adding the version as a separate col instead to make things clearer (and link the table header to the section)